posts/inside-rust/2022-08-11-diagnostic-effort.md

```diff
@@ -0,0 +1,323 @@
+            self.span,
+            fluent::parser::suggestion,
+            "->".to_string(),
+            Applicability::MachineApplicable,
+        )
+    }
+}
+```
+
+Instead of using strings for the messages as in the original diagnostic
+emission logic, typed identifiers referring to translation resources are used.
+Now we just have to [add the diagnostic messages to the translation
+resources..](#Adding-translation-resources…).
+
+#### Examples
+For more examples of diagnostics ported to use the diagnostic derive or written
+manually, see the following pull requests:
+
+- https://github.com/rust-lang/rust/pull/98353
+- https://github.com/rust-lang/rust/pull/98415
+- https://github.com/rust-lang/rust/pull/97093
+- https://github.com/rust-lang/rust/pull/99213
+
+There are more examples in [the complete list of diagnostic translation pull
+requests](https://hackmd.io/@davidtwco/rkXSbLg95).
+
+#### Adding translation resources...
+Every slug in a diagnostic derive or typed identifier in a manual
+implementation needs to correspond to a message in a translation resource.
+
+`rustc`'s translations use [Fluent][fluent], a asymmetric translation system.
+For each crate in the compiler which emits diagnostics, there is a
+corresponding Fluent resource at
+`compiler/rustc_error_messages/locales/en-US/$crate.ftl`.
+
+Error messages need to be added to this resource (a macro will then generate
+the typed identifier corresponding to the message).
+
+For our example, we should add the following Fluent to
+`compiler/rustc_error_messages/locales/en-US/parser.ftl`:
+
+```fluent
+parser-return-type-arrow = return types are denoted using `->`
+    .suggestion = use `->` instead
+```
+
+`parser-return-type-arrow` will generate a `parser::return_type_arrow` type (in
+`rustc_errors::fluent`) that can be used with diagnostic structs and the
+diagnostic builder.
+
+Subdiagnostics are "attributes" of the primary Fluent message - by convention,
+the name of attributes are the type of subdiagnostic, such as "suggestion", but
+this can be changed when there are multiple of one kind of subdiagnostic.
+
+Now that the Fluent resource contains the message, our diagnostic is ported!
+More complex messages with interpolation will be able to reference other fields
+in a diagnostic type (when implemented manually, those provided as arguments).
+See the diagnostic translation documentation [in the `rustc` dev
+guide][diag_translation] for more examples.
+
+### 3. Porting diagnostics
+Now that you've got a rough idea what to do, you need to find some diagnostics
+to port. There's lots of diagnostics to port, so the diagnostic working group
+have split the work up to avoid anyone working on the same diagnostic as
+someone else - but right now, there aren't many people involved, so just pick a
+crate and start porting it :)
+
+Please add any pull requests that you make [to this
+document](https://hackmd.io/@davidtwco/rkXSbLg95) so we can keep track of who
```

Comment:
We should make a tag instead of using this hackmd, I think.

Then the PR could be authored with something like `r? rust-lang/diagnostics @rustbot label +A-translation`  or something.

Reply:
I'll add one actually.

Reply:
https://github.com/rust-lang/rust/labels/A-translation

Reply:
Great, now we need to go add everything currently in that list to it :)

Reply:
> Great, now we need to go add everything currently in that list to it :)

Did that too.

Reply:
Great, updated this draft to reference it.